pkg/cloudprovider/aws/aws.go

```diff
@@ -601,3 +614,42 @@ func addASGTags(config *cloudprovider.NodeGroupConfig, asg *autoscaling.Group, p
 		log.Errorf("failed to create auto scaling tag for ASG %v", id)
 	}
 }
+
+// terminateOrphanedInstances will attempt to terminate a list of instances
+func terminateOrphanedInstances(n *NodeGroup, instances []*string) {
+	if len(instances) == 0 {
+		return
+	}
+	log.WithField("asg", n.id).Infof("terminating %v instance(s) that could not be attached to the ASG", len(instances))
+
+	var instanceIds []string
+	for i := 0; i < len(instances); i += terminateBatchSize {
+		batch := instances[i:minInt(i+terminateBatchSize, len(instances))]
+
+		for _, id := range batch {
+			instanceIds = append(instanceIds, *id)
+		}
+
+		_, err := n.provider.ec2Service.TerminateInstances(&ec2.TerminateInstancesInput{
+			InstanceIds: awsapi.StringSlice(instanceIds),
+		})
+		if err != nil {
+			log.Warnf("failed to terminate instances %v", err)
+		}
+	}
+
+	// prevent an endless cycle of provisioning and terminating instances
+	terminateInstancesTries++
+	if terminateInstancesTries == maxTerminateInstancesTries {
+		log.Fatalf("reached maximum number of consecutive failures (%v) for provisioning nodes with CreateFleet",
```

Comment:
Thoughts about `terminateInstancesTries` being a member of `NodeGroup` instead of Global? This fatal logic seems like it should be a NodeGroup fails 3 times in a row, rather than 3 NodeGroups fail once. I would like to avoid the situation of `Fatal`ing the process as much as possible, to avoid crash-looping for one bad nodegroup if the rest are okay.
 I'm okay Fatal being the action for now (@awprice ?), but per NodeGroup would allow this to be expanded to a per NodeGroup backoff or similar in the future.

cc @awprice

Reply:
Agree with @Jacobious52 - fatal is fine for now, but ideally this could be recoverable gracefully